File: lib/python/opentoken/src/main/opentoken/attributes/general/integer_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List, Optional
import re
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.validation import RegexValidator
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class IntegerAttribute(BaseAttribute):
    """Represents a generic integer attribute.

    This class extends BaseAttribute and provides functionality for working with
    integer fields. It recognizes "Integer" as a valid alias for this attribute type.

    The attribute performs normalization on input values by trimming whitespace
    and validates that the input is a valid integer.
    """

    NAME = "Integer"
    ALIASES = [NAME]

    # Regular expression pattern for validating integer format with optional sign
    VALIDATION_PATTERN = r"^\s*[+-]?\d+\s*$"

    def __init__(self, additional_validators: Optional[List[SerializableAttributeValidator]] = None):
        """
        Initialize the IntegerAttribute with optional additional validators.

        Args:
            additional_validators: Optional list of additional validators to apply
        """
        validation_rules: List[SerializableAttributeValidator] = [RegexValidator(self.VALIDATION_PATTERN)]
        if additional_validators:
            validation_rules.extend(additional_validators)
        super().__init__(validation_rules)

    def get_name(self) -> str:
        """Get the name of the attribute.

        Returns:
            str: The name "Integer"
        """
        return self.NAME

    def get_aliases(self) -> List[str]:
        """Get the aliases for the attribute.

        Returns:
            List[str]: A list containing the aliases for this attribute
        """
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalize the integer value by trimming whitespace and parsing.

        Args:
            value: The integer string to normalize

        Returns:
            str: The normalized integer value

        Raises:
            ValueError: If the integer is not valid or empty
        """
        if value is None:
            raise ValueError("Integer value cannot be null")
        if not value:
            raise ValueError("Invalid integer format: empty or whitespace")
            
        trimmed = value.strip()
        
        # Validate it's a valid integer
        try:
            int_value = int(trimmed)
            return str(int_value)
        except ValueError:
            raise ValueError(f"Invalid integer format: {value}")

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/general/IntegerAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import java.util.List;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.validation.RegexValidator;
import com.truveta.opentoken.attributes.validation.SerializableAttributeValidator;

/**
 * Represents a generic integer attribute.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * integer fields. It recognizes "Integer" as the primary alias for this attribute type.
 * 
 * The attribute performs normalization on input values, trimming whitespace and
 * ensuring the value is a valid integer.
 * 
 * The attribute also performs validation on input values, ensuring they are numeric
 * integers (no decimals, no negative sign handling beyond parsing).
 */
public class IntegerAttribute extends BaseAttribute {

    private static final String NAME = "Integer";
    private static final String[] ALIASES = new String[] { NAME };

    /**
     * Regular expression pattern for validating integer format.
     * 
     * This regex validates that the input is:
     * - Optional leading whitespace
     * - Optional leading sign (+ or -)
     * - One or more digits
     * - Optional trailing whitespace
     */
    private static final String INTEGER_REGEX = "^\\s*[+-]?\\d+\\s*$";

    public IntegerAttribute() {
        super(List.of(new RegexValidator(INTEGER_REGEX)));
    }

    /**
     * Protected constructor for subclasses to add additional validators.
     * 
     * @param additionalValidators additional validators to apply beyond the base integer format validation
     */
    protected IntegerAttribute(List<SerializableAttributeValidator> additionalValidators) {
        super(createValidatorList(additionalValidators));
    }

    private static List<SerializableAttributeValidator> createValidatorList(
            List<SerializableAttributeValidator> additionalValidators) {
        java.util.ArrayList<SerializableAttributeValidator> validators = new java.util.ArrayList<>();
        validators.add(new RegexValidator(INTEGER_REGEX));
        validators.addAll(additionalValidators);
        return validators;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        if (value == null) {
            throw new IllegalArgumentException("Integer value cannot be null");
        }

        // Trim whitespace and parse as integer
        String trimmed = value.trim();

        try {
            long intValue = Long.parseLong(trimmed);
            // Return the trimmed integer as a string
            return String.valueOf(intValue);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid integer format: " + value);
        }
    }

}
